### DISCUSSION TITLE
Lack of TimeStamp on EventEnvelope

### DISCUSSION THREAD

**Discussion by fabiomarreco** [akkadotnet/akka.net#4665] [Q&A] [answered]
Following Aaron's advice to try out this new discussion tool.

I am trying to find out how to extract the timestamp of the events from a persistence query. The class [EventEnvelope](https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Persistence.Query/EventEnvelope.cs) does not have this information in it.

The thing is, I've created a lot of events, and now I want to create a projection showing when a certain event occurred. When I look at the event journal table (in my case from oracle), I can see that there is a column in it as expected:

![](https://i.stack.imgur.com/o63gW.png)

So now I have 2 questions:

1) Is there a (not so hacky) way to retrive this information ? (I am currently doing a lookup on the event journal table that does not please me at all)

2) Why doesn't it have a timestamp in it ? I mean, all CQRS I've worked before (frameworks or when doing from scratch) had a timestamp on the event journal (and accessible on the projection). It was so intuitive to me that I never thought of adding it to the event in the first place. [Scala's implementation](https://doc.akka.io/japi/akka/current/akka/persistence/query/EventEnvelope.html) has a timestamp on the Envelope class.

Or maybe I'm just looking at the problem at the wrong angle, as I often do ;)

**Comment by Aaronontheweb** [comment]
This information is sometimes stored or derived as part of the `Offset` on the `EventEnevelope`: 

https://github.com/akkadotnet/akka.net/blob/77babf022eb7c06ecd8d12919af2441fed3b2fe1/src/core/Akka.Persistence.Query/Offset.cs#L50-L84

However, it looks to me like this is an oversight in the design of the `EventEnvelope`. There's no reason it shouldn't have an explicit timestamp appended to it given that this data is commonplace inside the Akka.Persistence journal.

We should create an issue and add some constructor overloads (can't modify the existing signatures - breaks binary compatibility) that time a Timestamp and exposed that as a public property on `EventEnvelope`.

**Reply by Aaronontheweb** [reply]
Created an issue for this here: https://github.com/akkadotnet/akka.net/issues/4668

**Reply by fabiomarreco** [reply]
Thanks Aaron! I´ll follow the Issue.
But I Kinda missed the point when you said this info is sometimes stored as part of the Offset. What do you mean by it ?

**Comment by Aaronontheweb** [comment, marked as answer]
This has now  been resolved: https://github.com/akkadotnet/akka.net/issues/4668
